### PR TITLE
AEA-1174 local registry sync

### PR DESCRIPTION
**Pull request by solarw**
## Proposed changes

local registry sync

## Fixes

https://github.com/fetchai/agents-aea/issues/2068, #1923

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [x] I have checked that the documentation about the `aea cli` tool works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: aea/cli/local_registry_sync.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+
+PACKAGES_DIRS = [i.to_plural() for i in PackageType]
+
+
+@click.command()
+@pass_ctx
+def local_registry_sync(ctx: Context):
+    """Upgrade the local package registry."""
+    do_local_registry_update(ctx.cwd)
```
Comment (DavidMinarsch) [inline aea/cli/local_registry_sync.py:44]: So this needs to be run from within `packages`?

File: aea/cli/local_registry_sync.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+    :return: generator with Tuple of package_id and package diurectory.
+    """
+    for author in os.listdir(base_dir):
+        author_dir = os.path.join(base_dir, author)
+        if not os.path.isdir(author_dir):
+            continue  # pragma: nocover
+        for package_type_plural in os.listdir(author_dir):
+            if package_type_plural not in PACKAGES_DIRS:
+                continue  # pragma: nocover
```
Comment (DavidMinarsch) [inline aea/cli/local_registry_sync.py:119]: We should raise an error when there are invalid directories.

File: aea/cli/local_registry_sync.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    Generate list of the packages in local repo directory.
+
+    :param base_dir: path or str of the local repo.
+
+    :return: generator with Tuple of package_id and package diurectory.
+    """
+    for author in os.listdir(base_dir):
+        author_dir = os.path.join(base_dir, author)
+        if not os.path.isdir(author_dir):
+            continue  # pragma: nocover
```
Comment (DavidMinarsch) [inline aea/cli/local_registry_sync.py:116]: We should raise an error when `packages` is of invalid format (i.e. invalid directories)

File: aea/cli/local_registry_sync.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        for package_type_plural in os.listdir(author_dir):
+            if package_type_plural not in PACKAGES_DIRS:
+                continue  # pragma: nocover
+            package_type_dir = os.path.join(author_dir, package_type_plural)
+            if not os.path.isdir(package_type_dir):
+                continue  # pragma: nocover
```
Comment (DavidMinarsch) [inline aea/cli/local_registry_sync.py:122]: see above

File: aea/cli/local_registry_sync.py (hunk 3 above)
Comment (DavidMinarsch) [reply aea/cli/local_registry_sync.py:116]: (unless the file is an init file, then we skip as you do)

File: aea/cli/local_registry_sync.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+            for package_name in os.listdir(package_type_dir):
+                package_dir = os.path.join(package_type_dir, package_name)
+                if not os.path.isdir(package_dir):
+                    continue  # pragma: nocover
```
Comment (DavidMinarsch) [inline aea/cli/local_registry_sync.py:126]: see above

File: aea/cli/local_registry_sync.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+                try:
+                    package_type = determine_package_type_for_directory(
+                        Path(package_dir)
+                    )
+                    if package_type.to_plural() != package_type_plural:
+                        # incorrect package placing
+                        continue  # pragma: nocover
```
Comment (DavidMinarsch) [inline aea/cli/local_registry_sync.py:133]: raise an error so user can fix it!

File: aea/cli/local_registry_sync.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+                    config = load_component_configuration(
+                        package_type, Path(package_dir), skip_consistency_check=True
+                    )
+                    yield (config.package_id, package_dir)
+                except ValueError:  # pragma: nocover
+                    # can not determine package type, not a package? just skip
+                    pass  # pragma: nocover
```
Comment (DavidMinarsch) [inline aea/cli/local_registry_sync.py:140]: I would raise, we want to have clean local registries

File: aea/cli/local_registry_sync.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+    """
+    for package_id, package_dir in enlist_packages(base_dir):
+        current_public_id = package_id.public_id
+        latest_public_id = get_package_latest_public_id(package_id)
+        if not (  # pylint: disable=superfluous-parens
+            current_public_id < latest_public_id
+        ):
+            click.echo(f"{package_id} is the latest version.")
+            continue
+        click.echo(f"Updating to {latest_public_id}.")
```
Comment (DavidMinarsch) [inline aea/cli/local_registry_sync.py:63]:
```suggestion
        click.echo(f"Updating {current_public_id} to {latest_public_id}.")
```

File: aea/cli/local_registry_sync.py [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+    :return: None
+    """
+    with TemporaryDirectory() as tmp_dir:
+        new_package_dir = os.path.join(tmp_dir, public_id.name)
+        os.mkdir(new_package_dir)
+        fetch_package(
+            package_type, public_id=public_id, cwd=tmp_dir, dest=new_package_dir
+        )
+        shutil.rmtree(package_dir)
+        shutil.move(new_package_dir, package_dir)
```
Comment (DavidMinarsch) [inline aea/cli/local_registry_sync.py:86]: If there is a failure here, we should handle it and keep the old package.

File: aea/configurations/loader.py [hunk 10]
```diff
@@ -308,18 +308,20 @@ def from_package_type(
 
 
 def load_component_configuration(
-    component_type: ComponentType,
+    component_type: Union[ComponentType, PackageType],
     directory: Path,
     skip_consistency_check: bool = False,
 ) -> "ComponentConfiguration":
```
Comment (DavidMinarsch) [inline aea/configurations/loader.py:314]: This is not ok imo. PackageType includes Agent, ComponentType does not. This is confusing terminology. You can create a new method which handles both, but this method should stay as is and be consistently named as is.

File: aea/configurations/loader.py [hunk 11]
```diff
@@ -329,7 +331,7 @@ def load_component_configuration(
 
 
 def _load_configuration_object(
-    component_type: ComponentType, directory: Path
+    package_type: PackageType, directory: Path
 ) -> ComponentConfiguration:
```
Comment (DavidMinarsch) [inline aea/configurations/loader.py:335]: note, you return a ComponentConfiguration but pass a package_type, that's inconcsistent!

File: aea/cli/local_registry_sync.py (hunk 9 above)
Comment (solarw) [reply aea/cli/local_registry_sync.py:86]: it does this logic fetch first, remove then

File: aea/cli/local_registry_sync.py (hunk 1 above)
Comment (solarw) [reply aea/cli/local_registry_sync.py:44]: yes

**Review by DavidMinarsch — APPROVED** [review]
LGTM

**Pull request merged**